File: protovalidate/validator.py
# Copyright 2023-2026 Buf Technologies, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from google.protobuf import message as google_message
from protobuf import Message, Registry

from protovalidate._gen.buf.validate import validate_pb
from protovalidate.internal import extra_func
from protovalidate.internal import rules as _rules
from protovalidate.internal.legacy import LegacyMessageConverter

CompilationError = _rules.CompilationError
Violations = validate_pb.Violations
Violation = _rules.Violation


class Validator:
    """
    Validates protobuf-py messages against static rules.

    Legacy google.protobuf messages are also accepted: they are copied into
    protobuf-py messages of the same type and validated identically. Violation
    field values then refer to the copy's values.

    Each validator instance caches internal state generated from the static
    rules, so reusing the same instance for multiple validations
    significantly improves performance.
    """

    _factory: _rules.RuleFactory

    def __init__(self, registry: Registry | None = None):
        """
        Parameters:
            registry: An optional protobuf-py Registry used to resolve custom
                predefined-rule extensions (proto2 extensions on the standard
                rule messages). Without it, only standard rules and rules whose
                extensions are known to the bundled stub are applied. For
                legacy google.protobuf messages, such a Registry can be built
                with ``FileDescriptorSet.to_registry()``.
        """
        funcs = extra_func.make_extra_funcs()
        self._factory = _rules.RuleFactory(funcs, registry)
        self._legacy = LegacyMessageConverter()

    def validate(self, message: Message | google_message.Message, *, fail_fast: bool = False):
        """
        Validates the given message against the static rules defined in
        the message's descriptor.

        Parameters:
            message: The message to validate.
            fail_fast: If true, validation will stop after the first iteration.
        Raises:
            CompilationError: If the static rules could not be compiled.
            ValidationError: If the message is invalid. The violations raised as part of this error should
            always be equal to the list of violations returned by `collect_violations`.
        """
        message = self._coerce(message)
        violations = self.collect_violations(message, fail_fast=fail_fast)
        if len(violations) > 0:
            msg = f"invalid {type(message).desc().name}"
            raise ValidationError(msg, violations)

    def collect_violations(
        self,
        message: Message | google_message.Message,
        *,
        fail_fast: bool = False,
    ) -> list[Violation]:
        """
        Validates the given message against the static rules defined in
        the message's descriptor. Compared to `validate`, `collect_violations` simply
        returns the violations as a list and puts the burden of raising an appropriate
        exception on the caller.

        The violations returned from this method should always be equal to the violations
        raised as part of the ValidationError in the call to `validate`.

        Parameters:
            message: The message to validate.
            fail_fast: If true, validation will stop after the first iteration.
        Raises:
            CompilationError: If the static rules could not be compiled.
        """
        message = self._coerce(message)
        ctx = _rules.RuleContext(fail_fast=fail_fast)
        for rule in self._factory.get(type(message).desc()):
            rule.validate(ctx, message)
            if ctx.done:
                break
        for violation in ctx.violations:
            violation.finalize_paths()
        return ctx.violations

    def _coerce(self, message: Message | google_message.Message) -> Message:
        if isinstance(message, google_message.Message):
            return self._legacy.convert(message)
        return message


class ValidationError(ValueError):
    """
    An error raised when a message fails to validate.
    """

    _violations: list[_rules.Violation]

    def __init__(self, msg: str, violations: list[_rules.Violation]):
        super().__init__(msg)
        self._violations = violations

    def to_proto(self) -> validate_pb.Violations:
        """
        Provides the Protobuf form of the validation errors.
        """
        return validate_pb.Violations(violations=[violation.proto for violation in self._violations])

    @property
    def violations(self) -> list[Violation]:
        """
        Provides the validation errors as a simple Python list, rather than the
        Protobuf-specific collection type used by Violations.
        """
        return self._violations

File: crates/protovalidate/src/validator.rs
// Copyright (c) 2023-2026 Buf Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::collections::HashMap;
use std::fmt;
use std::sync::{Arc, PoisonError, RwLock, RwLockReadGuard};

use buffa::Message as _;
use buffa_descriptor::MessageIndex;
use buffa_descriptor::generated::descriptor::FileDescriptorSet;

#[cfg(feature = "cel")]
use crate::cel;
use crate::descriptors::{self, Descriptors};
use crate::error::Internal;
use crate::protobuf::{Payload, Runtime};
use crate::rules::MessageEvaluator;
use crate::rules::build::Builder;
use crate::rules::eval::Walker;
use crate::validate::{Violation as ViolationPb, Violations};
use crate::{DescriptorError, Error, ValidationError};

/// The compiled rules of every message type validated so far.
type Cache = HashMap<MessageIndex, Arc<MessageEvaluator>>;

/// The violations as a serialized `buf.validate.Violations`.
fn encode_violations(violations: Vec<ViolationPb>) -> Vec<u8> {
    Violations {
        violations,
        ..Default::default()
    }
    .encode_to_vec()
}

/// A CEL environment knowing the `buf.validate` schema.
#[cfg(feature = "cel")]
fn cel_env() -> cel::Env {
    let mut env = cel::new_env()
        .unwrap_or_else(|error| panic!("could not initialize the CEL runtime: {error}"));
    let set = descriptors::decode_file_set(descriptors::BASE_DESCRIPTOR_SET)
        .expect("embedded descriptor set decodes");
    for file in &set.file {
        env.add_file(&file.encode_to_vec())
            .unwrap_or_else(|error| panic!("could not register the buf.validate schema: {error}"));
    }
    env
}

/// Validates Protobuf messages against the rules in their descriptors.
///
/// Descriptors are registered serialized, and a message is read in place
/// through a [`Runtime`] the caller implements over its own Protobuf
/// runtime.
pub struct Validator {
    descriptors: Descriptors,
    #[cfg(feature = "cel")]
    env: RwLock<cel::Env>,
    cache: RwLock<Cache>,
}

impl fmt::Debug for Validator {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Validator").finish_non_exhaustive()
    }
}

impl Default for Validator {
    fn default() -> Self {
        Self::new()
    }
}

impl Validator {
    /// Creates a validator.
    ///
    /// Use `add_*` methods to register file descriptors for the messages you will validate.
    #[must_use]
    pub fn new() -> Self {
        Self {
            descriptors: Descriptors::new(),
            #[cfg(feature = "cel")]
            env: RwLock::new(cel_env()),
            cache: RwLock::new(HashMap::new()),
        }
    }

    /// Registers a serialized `google.protobuf.FileDescriptorProto`.
    ///
    /// A file's imports must be registered before the file itself. Adding a
    /// file whose name is already known is a no-op, so descriptors
    /// may be re-added freely.
    ///
    /// # Errors
    ///
    /// Fails if the bytes do not parse, the file is invalid, or one of its
    /// imports has not been registered.
    pub fn add_file_descriptor_bytes(&mut self, file: &[u8]) -> Result<(), DescriptorError> {
        let proto = descriptors::decode_file(file).map_err(DescriptorError::new)?;
        let set = FileDescriptorSet {
            file: vec![proto],
            ..Default::default()
        };
        self.descriptors
            .add_file_set(set)
            .map_err(DescriptorError::new)?;
        #[cfg(feature = "cel")]
        self.env
            .get_mut()
            .unwrap_or_else(PoisonError::into_inner)
            .add_file(file)
            .map_err(|error| DescriptorError::new(error.to_string()))?;
        Ok(())
    }

    /// Validates a message read in place through a [`Runtime`].
    ///
    /// `type_name` is the fully-qualified message type, without a leading
    /// dot; its descriptor must already be registered. `payload` is the
    /// message's serialized form, produced only if a CEL rule binds the
    /// message itself, a repeated field or a map to `this`. With `fail_fast`
    /// validation stops at the first violation, rather than accumulating
    /// them all.
    ///
    /// # Errors
    ///
    /// A message that breaks its rules returns [`Error::Validation`],
    /// carrying the violations as a serialized `buf.validate.Violations`.
    /// The other variants mean validation itself failed: an unknown type or
    /// unparsable payload ([`Error::Argument`]), rules that do not compile
    /// ([`Error::Compilation`]), or a rule failing to evaluate
    /// ([`Error::Evaluation`]).
    pub fn validate_message<R: Runtime>(
        &self,
        type_name: &str,
        message: &R::Message<'_>,
        payload: Payload<'_>,
        fail_fast: bool,
    ) -> Result<(), Error> {
        let violations = self.run::<R>(type_name, message, payload, fail_fast)?;
        if violations.is_empty() {
            return Ok(());
        }
        Err(Error::Validation(ValidationError::new(encode_violations(
            violations,
        ))))
    }

    fn run<R: Runtime>(
        &self,
        type_name: &str,
        message: &R::Message<'_>,
        payload: Payload<'_>,
        fail_fast: bool,
    ) -> Result<Vec<ViolationPb>, Internal> {
        let index = self.message_index(type_name)?;
        let evaluators = self.evaluators(index)?;
        #[cfg(feature = "cel")]
        let env = self.env.read().unwrap_or_else(PoisonError::into_inner);
        #[cfg(feature = "cel")]
        let walker = Walker::<R>::new(&self.descriptors, &env, &evaluators, fail_fast);
        #[cfg(not(feature = "cel"))]
        let walker = Walker::<R>::new(&self.descriptors, &evaluators, fail_fast);
        Ok(walker.validate(message, type_name, payload, &evaluators[&index])?)
    }

    fn message_index(&self, type_name: &str) -> Result<MessageIndex, Internal> {
        self.descriptors
            .pool
            .message_index(type_name)
            .ok_or_else(|| Internal::Argument(format!("unknown message type: {type_name}")))
    }

    /// The cache, with the rules of `index` and every type reachable from
    /// it compiled.
    ///
    /// Rules compile lazily, on the first validation of a type, the whole
    /// reachable closure at once. Failures are not cached and are reported
    /// again on every call.
    fn evaluators(&self, index: MessageIndex) -> Result<RwLockReadGuard<'_, Cache>, Internal> {
        let cache = self.read_cache();
        if cache.contains_key(&index) {
            return Ok(cache);
        }
        let mut built = HashMap::new();
        #[cfg(feature = "cel")]
        {
            let mut env = self.env.write().unwrap_or_else(PoisonError::into_inner);
            let mut builder = Builder::new(&self.descriptors, &mut env);
            builder.build_closure(index, &cache, &mut built)?;
        }
        #[cfg(not(feature = "cel"))]
        {
            let mut builder = Builder::new(&self.descriptors);
            builder.build_closure(index, &cache, &mut built)?;
        }
        drop(cache);
        self.cache
            .write()
            .unwrap_or_else(PoisonError::into_inner)
            .extend(built);
        Ok(self.read_cache())
    }

    fn read_cache(&self) -> RwLockReadGuard<'_, Cache> {
        self.cache.read().unwrap_or_else(PoisonError::into_inner)
    }
}
